From cd1b0c295dc4ef2ce5e6113798c340ec4785ca14 Mon Sep 17 00:00:00 2001
From: Gerd Hoffmann <kraxel@redhat.com>
Date: Wed, 7 Mar 2012 08:00:20 +0100
Subject: [PATCH 06/12] suspend: make serial ports wakeup the guest.

RH-Author: Gerd Hoffmann <kraxel@redhat.com>
Message-id: <1331107226-21901-7-git-send-email-kraxel@redhat.com>
Patchwork-id: 38356
O-Subject: [RHEL-6.3 qemu-kvm PATCH v3 06/12] suspend: make serial ports wakeup the guest.
Bugzilla: 766303
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>
RH-Acked-by: Gleb Natapov <gleb@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>

Add a 'wakeup' property to the serial port.  It is off by default.  When
enabled any incoming character on the serial line will wake up the
guest.  Useful for guests which have a serial console configured.

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>

upstream: 9826fd597df59a8bac7beafd192e4baad790c31a
---
 hw/serial.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

Signed-off-by: Michal Novotny <minovotn@redhat.com>
---
 hw/serial.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/hw/serial.c b/hw/serial.c
index 29151e9..8e92f67 100644
--- a/hw/serial.c
+++ b/hw/serial.c
@@ -131,6 +131,7 @@ struct SerialState {
     int it_shift;
     int baudbase;
     int tsr_retry;
+    uint32_t wakeup;
 
     uint64_t last_xmit_ts;              /* Time when the last byte was successfully sent out of the tsr */
     SerialFIFO recv_fifo;
@@ -621,6 +622,10 @@ static int serial_can_receive1(void *opaque)
 static void serial_receive1(void *opaque, const uint8_t *buf, int size)
 {
     SerialState *s = opaque;
+
+    if (s->wakeup) {
+        qemu_system_wakeup_request(QEMU_WAKEUP_REASON_OTHER);
+    }
     if(s->fcr & UART_FCR_FE) {
         int i;
         for (i = 0; i < size; i++) {
@@ -911,6 +916,7 @@ static ISADeviceInfo serial_isa_info = {
         DEFINE_PROP_HEX32("iobase", ISASerialState, iobase,  -1),
         DEFINE_PROP_UINT32("irq",   ISASerialState, isairq,  -1),
         DEFINE_PROP_CHR("chardev",  ISASerialState, state.chr),
+        DEFINE_PROP_UINT32("wakeup", ISASerialState, state.wakeup, 0),
         DEFINE_PROP_END_OF_LIST(),
     },
 };
-- 
1.7.7.6

